### PR TITLE
RFC: TRL backend for Arctic RL

### PR DESCRIPTION
**Pull request by kashif**
Sketch of a TRL backend for Arctic RL, opened as an RFC alongside huggingface/trl#6676. Not ready to merge: the batch layout helpers are stubs and none of this has run against a live server.

The idea is that TRL keeps the loss. `forward` returns per token logprobs, TRL builds its loss on them, and `backward` sends back `d(loss)/d(logprobs)`. The server backprops `sum(w * logprobs)`, which gives the same parameter gradients as the real loss. Verified bit exact against the in process path on TRL's GRPO loss.

So one server side loss function instead of one per algorithm. For comparison, `integrations/verl/grpo_loss.py` is 523 lines and still covers `vanilla` only, out of the 12 in verl's `POLICY_LOSS_REGISTRY`. `integrations/trl/loss.py` is about 6 lines and would not need touching when TRL adds a loss variant.

Two things I could not settle from the code:

1. `pipeline._resolve_fn` falls back to a dotted path import, so on a self hosted server the loss resolves without a registry entry at all. Is that a supported extension point or an implementation detail?
2. For managed deployments, would a generic registry entry be ok? Weighted cross entropy is the same surrogate under `weights = -d(loss)/d(logprobs)`, which is how Tinker avoids adding a loss.

Remaining open items are in the README. The main one is TRL side: it still wants a local model to prepare and to stream weights from.

### CODE REVIEW

File: arctic_platform/integrations/trl/client.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    ) -> ForwardOutput:
+        # `model` is the trainer's local module. Arctic owns the weights, so it is unused here.
+        #
+        # AsyncGRPOTrainer packs a rank's samples into one padding-free row and marks sequence boundaries by
+        # position_ids resets. The server wants padded [B, S] rows, so unpack first. This conversion is the
+        # bulk of the adapter and is the main reason it belongs on this side of the wire.
+        batch = _unpack_to_padded_rows(input_ids, position_ids, completion_mask)
+        self._batch = batch
+
+        response = self.client.fwd_no_grad(
```
Comment (sfc-gh-sbekman) [inline arctic_platform/integrations/trl/client.py:83]: how would bwd work if fwd is no_grad?

I think this `forward` method should contain no_grad -> `forward_no_grad` - otherwise the API is confusing, no?

File: arctic_platform/integrations/trl/client.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        #
+        # The server returns log probs under the roll(-1) convention and TRL expects the [:, 1:] shift, so
+        # the repack aligns them rather than merely reshaping.
+        return ForwardOutput(
+            log_probs=_repack_to_row(out["logprobs"], completion_mask).requires_grad_(True),
+            entropy=_repack_to_row(out["entropy"], completion_mask),
+            aux_loss=None,  # Arctic reports the MoE aux loss as a metric, not as a differentiable tensor
+        )
+
+    def backward(self, grad_log_probs: torch.Tensor) -> None:
```
Comment (sfc-gh-sbekman) [inline arctic_platform/integrations/trl/client.py]: same here - shouldn't the API say it's `forward_backward`?

**Review by sfc-gh-sbekman — COMMENTED** [review]
You're fast, Kashif!

I have some questions:

File: arctic_platform/integrations/trl/client.py (hunk 1 above)
Comment (kashif) [reply arctic_platform/integrations/trl/client.py:83]: Renamed, here and in the TRL protocol, so it is `forward_no_grad` and `forward_backward` on both sides now.

On how bwd works: `fwd_bwd` runs its own forward on the server, so the graph is rebuilt there and the backward is real. The scoring pass cannot carry a graph across the wire, which is the only reason there are two forwards. Added a comment at both call sites since that was not obvious from the code.

File: arctic_platform/integrations/trl/client.py (hunk 2 above)
Comment (kashif) [reply arctic_platform/integrations/trl/client.py]: Done, `forward_backward` now.

File: arctic_platform/integrations/trl/client.py (hunk 1 above)
Comment (sfc-gh-sbekman) [reply arctic_platform/integrations/trl/client.py:83]: thank you, Kashif

looks good to me now

Are you planning to merge this already or do more work on this - wasn't sure if to approve it or wait

File: arctic_platform/integrations/trl/README.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+## What TRL changes
+
+Nothing. `training_client=` already exists on `AsyncGRPOTrainer` alongside the `rollout_worker` and
+`weight_transfer` hooks, and `transformers.Trainer` already accepts `optimizers=`, which is why there is no
+`optim_step` or `clip_grad_norm` in the protocol.
+
+## Open items
+
+1. **No local model.** The real blocker. Arctic owns the weights, but transformers and accelerate still want
```
Comment (sfc-gh-truwase) [inline arctic_platform/integrations/trl/README.md:71]: @kashif, yes this is intentional and a core design principle. The components (trainer, rollout, ..) reside on the server side. For example, the trainer uses [transformers](https://github.com/Snowflake-AI-Research/Arctic-Platform/blob/9e748ec3ebdb88f663066540c41b81368b7d4405/arctic_platform/common/deepspeed_worker.py#L163) for model preparation and [liger](https://github.com/Snowflake-AI-Research/Arctic-Platform/blob/9e748ec3ebdb88f663066540c41b81368b7d4405/arctic_platform/common/deepspeed_worker.py#L173-L175) for patching. This allows TRL to run on CPU-only machines. Would this work?

File: arctic_platform/integrations/trl/README.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+   a local module to `prepare()`, and `AsyncGRPOTrainer._sync_weight` streams from
+   `self.model.named_parameters()` on every rank (`async_grpo_trainer.py:1149`). Meta device crashes there
+   and an empty CUDA model wastes full model memory. A stub module carrying names and shapes is the likely
+   answer. TRL-side work.
+2. **Microbatch scaling.** The adapter folds the trainer's gradient-accumulation scaling into the weights
```
Comment (sfc-gh-truwase) [inline arctic_platform/integrations/trl/README.md:76]: This is an important point that I believe should be handled by an explicit contract: server-side vs client-side scaling.

**Comment by sfc-gh-sbekman** [comment]
Kashif, one important thing we need to discuss is that in order to do fwd(server) + loss(client) + fwd/bwd(server) approach correctly - we must replay the 2nd forward exactly the same  as the first one - which means, the same RNG engine and if it's an MoE model we also need a router replay for the 2nd fwd.

This is our issue to solve, but I'm just bringing it up that w/o it that Tinker algo won't work correctly

**Comment by kashif** [comment]
Superseded by #84, which carries these three commits rebased and takes them to a validated end to end run, so closing this to keep the review in one place.

@sfc-gh-sbekman that also answers your question above about whether to approve or wait: this one was never going to merge on its own, the batch layout helpers were stubs.
